File: 0310-minimum-height-trees/0310-minimum-height-trees.java
class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        
        if (n == 1) {
            return Collections.singletonList(0);
        }
        
        int[] inEdges = new int[n];
        ArrayList<ArrayList<Integer>> adj = new ArrayList<>(n);
        
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }
        
        for (int i = 0; i < edges.length; i++) {
            int u = edges[i][0], v = edges[i][1];
            inEdges[u]++;
            inEdges[v]++;
            adj.get(u).add(v);
            adj.get(v).add(u);
        }
        
        Queue<Integer> q = new LinkedList<>();
        
        for (int i = 0; i < n; i++) {
            if (inEdges[i] == 1) {
                q.add(i);
            }
        }
        
        List<Integer> ans = new ArrayList<>();
        
        while (!q.isEmpty()) {
            int sz = q.size();
            List<Integer> res = new ArrayList<>();
            while (sz-- > 0) {
                int curr = q.poll();
                res.add(curr);
                
                for (int neigh : adj.get(curr)) {
                    inEdges[neigh]--;
                    
                    if (inEdges[neigh] == 1) {
                        q.add(neigh);
                    }
                }
            }
            
            ans = res;
        }
        
        return ans;
    }
}

File: 2025-and-before/0310-minimum-height-trees/0310-minimum-height-trees.cpp
class Solution {
public:
    vector<int> findMinHeightTrees(int n, vector<vector<int>>& edges) {
        
        vector<vector<int>>adj(n);
        vector<int>inEdge(n,0);
        
        for(vector<int>&edg : edges){
            
            inEdge[edg[0]]++;
            inEdge[edg[1]]++;
            
            adj[edg[0]].push_back(edg[1]);
            adj[edg[1]].push_back(edg[0]);
        }
        
        
        queue<int>q;
        for(int i=0;i<n;i++){
            
            if(inEdge[i] == 1)
                q.push(i);
        }
        
        vector<int>ans;
        
        while(q.size()){
            
            int sz = q.size();
            ans.clear();
            
            while(sz--){
                
                int curr = q.front();
                q.pop();
                
                ans.push_back(curr);
               
                for(int&neigh : adj[curr]){
                    inEdge[neigh]--;
                    
                    if(inEdge[neigh] == 1){
                        q.push(neigh);
                    }
                }       
            }
        }
        
        if(ans.size() == 0)
            ans.push_back(0);
        
        
        return ans;
        
    }
};